collector/kubernetes.go

```diff
@@ -0,0 +1,133 @@
+	errors  *prometheus.CounterVec
+	client  *godo.Client
+	timeout time.Duration
+
+	Up        *prometheus.Desc
+	Count     *prometheus.Desc
+	NodePools *prometheus.Desc
+	Nodes     *prometheus.Desc
+}
+
+// NewKubernetesCollector returns a new KubernetesCollector
+func NewKubernetesCollector(logger log.Logger, errors *prometheus.CounterVec, client *godo.Client, timeout time.Duration) *KubernetesCollector {
+	errors.WithLabelValues("kubernetes").Add(0)
+
+	// Version refers to the upstream Kubernetes version as well as the DigitalOcean revision
+	clusterLabels := []string{"id", "name", "region", "version"}
+	nodeLabels := []string{"id", "name", "region"}
+	return &KubernetesCollector{
+		logger:  logger,
+		errors:  errors,
+		client:  client,
+		timeout: timeout,
+
+		Count: prometheus.NewDesc(
+			"digitalocean_kubernetes_cluster_count",
+			"Number of Kubernetes clusters",
+			nil, nil,
+		),
+		Up: prometheus.NewDesc(
+			"digitalocean_kubernetes_cluster_up",
+			"If 1 the kubernetes cluster is up and running, 0 otherwise",
+			clusterLabels, nil,
+		),
+		NodePools: prometheus.NewDesc(
+			"digitalocean_kubernetes_nodepools_count",
+			"Number of Kubernetes nodepools",
+			clusterLabels, nil,
+		),
+		Nodes: prometheus.NewDesc(
+			"digitalocean_kubernetes_nodes_count",
+			"Number of Kubernetes nodes",
+			nodeLabels, nil,
+		),
+	}
+}
+
+// Describe secnds the super-set of all possible descriptors of metrics collected by this Collector.
+func (c *KubernetesCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- c.Count
+	ch <- c.Up
+	ch <- c.NodePools
+}
+
+// Collect is called by the Prometheus registry when collecting metrics
+func (c *KubernetesCollector) Collect(ch chan<- prometheus.Metric) {
+	ctx, cancel := context.WithTimeout(context.Background(), c.timeout)
+	defer cancel()
+	clusters, _, err := c.client.Kubernetes.List(ctx, nil)
+	if err != nil {
+		c.errors.WithLabelValues("kubernetes").Add(1)
+		level.Warn(c.logger).Log(
+			"msg", "can't list clusters",
+			"err", err,
+		)
+	}
+
+	ch <- prometheus.MustNewConstMetric(
+		c.Count,
+		prometheus.GaugeValue,
+		float64(len(clusters)),
+		[]string{}...,
+	)
+
+	for _, cluster := range clusters {
+		labels := []string{
+			cluster.ID,
+			cluster.Name,
+			cluster.RegionSlug,
+			cluster.VersionSlug,
+		}
+
+		var active float64
+		//TODO(dazwilkin) better reflect richer Kubernetes cluster states
+		if cluster.Status.State == godo.KubernetesClusterStatusRunning {
+			active = 1.0
+		}
+		ch <- prometheus.MustNewConstMetric(
+			c.Up,
+			prometheus.GaugeValue,
+			active,
+			labels...,
+		)
+		ch <- prometheus.MustNewConstMetric(
+			c.NodePools,
+			prometheus.GaugeValue,
+			float64(len(cluster.NodePools)),
+			labels...,
+		)
+
+		for _, nodepool := range cluster.NodePools {
+			// Assume NodePools are constrained to the cluster's Region
+			// If so, we can labels a cluster's NodePools by the cluster's region
+			go func(region string, np *godo.KubernetesNodePool) {
```

Comment:
I need to double check what the underlying implementation is and if this channel is buffered. But s
Nevertheless if we don't use the wait group we also can't use a goroutine here. 
I'm reviewing this on my phone, so I think it should work without the goroutine, but I need to verify that later.

Reply:
Crap, sorry. My mistake.

Resolved: https://github.com/metalmatze/digitalocean_exporter/pull/5/commits/b9b69c3de21556352cb0ea8900dcec95b7b39152